packages/gatsby/src/redux/nodes.js

```diff
@@ -103,6 +103,34 @@ exports.saveResolvedNodes = async (nodeTypeNames, resolver) => {
   }
 }
 
+/**
+ * Get node and save path dependency.
+ *
+ * @param {string} typeName
+ * @param {string} id
+ * @returns {Object|void} node
+ */
+const getResolvedNode = (typeName, id) => {
+  const { nodesByType, resolvedNodesCache } = store.getState()
+  const nodes /*: Map<mixed> */ = nodesByType.get(typeName)
+
+  if (!nodes) return undefined
+
+  let node = nodes.get(id)
+
+  if (!node) return undefined
+
+  const resolvedNodes = resolvedNodesCache.get(typeName)
+
+  if (resolvedNodes) {
+    node.__gatsby_resolved = resolvedNodes.get(id)
+  }
+
+  return node
+}
```

Comment:
This function looks a lot like `addResolvedNodes`. I guess it is a special case of it so maybe we could re-use common logic of those two.

Reply:
The important point is that it doesn't iterate the entire list of nodes. That's very important when scaling up. And since `nodes` is a Map, there's no short circuit mechanism like `.some` or `.every`. Any short cicuiting approach would mean doing a full loop through all elements, or worse (generate an array with all keys or smth, before being able to do a partial loop).

If we were to track a shadow array with all keys of the map then short circuiting would be an option. This would cut down mandatory Map induced O(n) or O(n^2) loops by half. Not relevant for big oh, but very relevant if the total runtime is "only" 2h. But that's for another time.

There's really just three steps in the new function; get nodes by type, fetch node by index, decorate node. The existing function does the first and, in a loop, the second.

While it's possible to share code, I'm not seeing a solution where the solution isn't at least as bad-if-not-worse in terms of maintanance than the existing duplication. I am open to suggestions.